File: app/src/main/java/com/example/mg/todo/data/DataProvider.java
package com.example.mg.todo.data;

import com.example.mg.todo.data.model.NoteModel;
import com.example.mg.todo.ui.NotesActivity.DI.INoteActivityScope;
import com.example.mg.todo.ui.NotesActivity.INotesContract;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

@INoteActivityScope
public class DataProvider implements INotesContract.IDate {
    private PrefrenceHelper prefrenceHelper;
    private List<NoteModel> mNoteModels;

    @Inject
    DataProvider(PrefrenceHelper helper) {
        prefrenceHelper = helper;
    }

    @Override
    public void updateDataSet() {
        prefrenceHelper.saveData(mNoteModels);
    }

    @Override
    public void addNote(NoteModel newNote) {
        mNoteModels.add(newNote);
        updateDataSet();
    }

    @Override
    public void updateNote(NoteModel newNote, int mUpdated) {
        mNoteModels.set(mUpdated, newNote);
        updateDataSet();
    }

    @Override
    public NoteModel getNote(int position) {
        return mNoteModels.get(position);
    }

    @Override
    public List<NoteModel> getDataModels() {
        mNoteModels = prefrenceHelper.getSavedData() != null ?
                prefrenceHelper.getSavedData() : new ArrayList<>();
        return mNoteModels;
    }
}

File: app/src/main/java/com/example/mg/todo/data/DataProvider.kt
package com.example.mg.todo.data

import android.app.Application
import com.example.mg.todo.data.model.NoteModel

class DataProvider(application: Application) {

    private val databaseHelper: INotesDao = DatabaseHelper.invoke(application).notesDao()

    suspend fun addNote(model: NoteModel): Long {
        return databaseHelper.insertNote(model)
    }

    suspend fun updateNote(model: NoteModel): Int {
        return databaseHelper.updateNote(model)
    }

    suspend fun removeNote(id: NoteModel): Int {
        return databaseHelper.deleteNote(id)
    }

    suspend fun dataModels(): List<NoteModel> {
        return databaseHelper.getAllNotes()
    }

}